# Enforce console logging for useAlertDialog by severity (`@blumintinc/blumint/enforce-console-error`)

üíº This rule is enabled in the ‚úÖ `recommended` config.

<!-- end auto-generated rule header -->

## Rule Details

Enforce proper logging for useAlertDialog based on severity. When severity is "error", console.error must be included. When severity is "warning", console.warn must be included. This ensures all user-facing errors and warnings are properly logged to monitoring systems.

- severity: `"error"` ‚Üí must call `console.error(...)` in the same function scope
- severity: `"warning"` ‚Üí must call `console.warn(...)` in the same function scope

### ‚ùå Incorrect

```ts
function submit() {
  useAlertDialog({ severity: 'error', message: 'Failed' }); // missing console.error
}
```

### ‚úÖ Correct

```ts
function submit(userId: string) {
  console.error('submit failed', { userId, context: 'submit' });
  useAlertDialog({ severity: 'error', message: 'Failed' });
}
```

## Options

None.

## When Not To Use It

Tests or code paths where logging is intentionally centralized elsewhere.
