# Enforce unique cursor headers (`@blumintinc/blumint/enforce-unique-cursor-headers`)

<!-- end auto-generated rule header -->

Ensures each checked file has exactly one cursor header before any code, preventing missing metadata and duplicate header blocks that create conflicting ownership information. The rule is designed for the standard cursor header that includes `@fileoverview`, but it can be configured to require additional tags, restrict the files that need headers, and optionally insert a default header template when one is missing.

## Rule Details

- Flags files that match `requiredPatterns` but have no cursor header containing all `requiredTags` ahead of the first import or statement.
- Flags multiple cursor headers before code; by default adjacent header fragments can be merged, but setting `allowSplitHeaders: false` requires metadata to live in a single block.
- Respects `excludedPatterns` and skips files marked with generated markers (e.g., `auto-generated`, `@generated`) when `ignoreGeneratedFiles` is true.
- Autofix removes duplicate headers and, when `headerTemplate` is provided, inserts a header at the top of the file.

### Options

```json
{
  "requiredPatterns": ["**/*.{ts,tsx,js,jsx}"],
  "excludedPatterns": ["**/*.d.ts", "**/dist/**", "**/build/**", "**/coverage/**"],
  "requiredTags": ["@fileoverview"],
  "allowSplitHeaders": true,
  "headerTemplate": null,
  "ignoreGeneratedFiles": true,
  "generatedMarkers": ["@generated", "auto-generated", "autogenerated", "do not edit", "do not edit manually"]
}
```

- `requiredPatterns`: Globs of files that must contain a cursor header.
- `excludedPatterns`: Globs to skip (e.g., definition files or build outputs).
- `requiredTags`: Tags that must appear in the header text.
- Header detection relies on `requiredTags` and `@`-prefixed comment lines; no tag (including `@fileoverview`) is hardcoded beyond the defaults you configure.
- `allowSplitHeaders`: When false, metadata spread across multiple adjacent comment blocks is reported as a duplicate header.
- `headerTemplate`: Optional comment block to insert when a required header is missing.
- When `headerTemplate` already ends with newline characters, the fixer preserves them and only adds the minimum extra whitespace to leave a clear blank line before code.
- `ignoreGeneratedFiles`: When true, generated files identified by `generatedMarkers` are skipped.
- `generatedMarkers`: Strings that signal the file is generated.

## Examples

Examples of **incorrect** code for this rule:

```ts
// No header at all
import { HttpsError } from '../errors/HttpsError';
export const validateUser = (userId: string) => !!userId;
```

```ts
/**
 * @fileoverview User authentication utilities
 */

/**
 * @fileoverview User authentication utilities
 */
export const validateUser = (userId: string) => !!userId;
```

```ts
import { utils } from './utils';

/**
 * @fileoverview Header after imports is ignored
 */
export const lateHeader = () => utils;
```

Examples of **correct** code for this rule:

```ts
/**
 * @fileoverview User authentication utilities
 * @author BluMint
 */
import { HttpsError } from '../errors/HttpsError';
export const validateUser = (userId: string) => !!userId;
```

```ts
// allowSplitHeaders = true (default)
/**
 * @fileoverview User authentication utilities
 */
/**
 * @author BluMint
 */
export const helper = () => {};
```

```ts
// allowSplitHeaders = false â†’ requires single block
/**
 * @fileoverview User authentication utilities
 * @author BluMint
 */
export const helper = () => {};
```

## When Not To Use It

Disable this rule if your project does not require cursor headers, or if another tool guarantees file-level metadata and you want to allow multiple header blocks or omit `@fileoverview`.
