# Ensure files have exactly one cursor header containing required tags (e.g., @fileoverview) before any code (`@blumintinc/blumint/enforce-unique-cursor-headers`)

ðŸ’¼ This rule is enabled in the âœ… `recommended` config.

ðŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

Ensure each checked file has exactly one cursor header before any code so ownership metadata stays discoverable. You can require additional tags, control which files need headers, and optionally insert a default header template when one is missing.

## Rule Details

- Add a cursor header containing all `requiredTags` to any file that matches `requiredPatterns` before the first import or statement.
- Keep exactly one cursor header before code; when `allowSplitHeaders` is true you can keep adjacent fragments that together contain the required tags, otherwise keep the metadata in a single block.
- Use `excludedPatterns` to skip files and rely on `ignoreGeneratedFiles` plus `generatedMarkers` to avoid touching generated sources.
- Autofix removes duplicate headers that already include all required tags and inserts `headerTemplate` when provided. The `headerTemplate` MUST contain all `requiredTags` to be inserted; if it is empty or missing tags, the rule will report the error but will not suggest an auto-fix to avoid infinite fix loops. With `allowSplitHeaders: false`, adjacent fragments are reported without auto-fix so you can merge them without losing optional metadata; skip `--fix` if you intentionally keep split blocks or non-required tags.

### Options

```json
{
  "requiredPatterns": ["**/*.{ts,tsx,js,jsx}"],
  "excludedPatterns": ["**/*.d.ts", "**/dist/**", "**/build/**", "**/coverage/**"],
  "requiredTags": ["@fileoverview"],
  "allowSplitHeaders": true,
  "headerTemplate": null,
  "ignoreGeneratedFiles": true,
  "generatedMarkers": ["@generated", "auto-generated", "autogenerated", "do not edit", "do not edit manually"],
  "excludedAtDirectives": ["@ts-nocheck", "@ts-check", "@ts-ignore", "@ts-expect-error", "@jsx", "@jsxRuntime", "@jsxImportSource", "@format", "@flow", "@jest-environment", "@vitest-environment"]
}
```

- `requiredPatterns`: Specify the globs of files where you require a cursor header.
- `excludedPatterns`: Skip files (definition files, build outputs, or other generated artifacts).
- `requiredTags`: List the tags that must appear in the header text.
- Header detection relies on `requiredTags` and `@`-prefixed comment lines that are not listed in `excludedAtDirectives` (for example `@ts-nocheck`, `@format`, or `@jest-environment`), so your top-of-file pragmas are not merged into the cursor header.
- `allowSplitHeaders`: When false, metadata spread across multiple adjacent comment blocks is reported as a duplicate header, but the fixer does not delete those fragments. You must manually merge them into a single block; use caution with `--fix` if you intentionally keep separate metadata sections.
- `headerTemplate`: Provide a comment block to insert when a required header is missing. The template MUST contain all `requiredTags` to be applied via auto-fix.
- When `headerTemplate` already ends with newline characters, the fixer preserves them and only adds the minimum extra whitespace to leave a clear blank line before code.
- `ignoreGeneratedFiles`: When true, generated files identified by `generatedMarkers` are skipped.
- `generatedMarkers`: Identify strings that signal the file is generated.
- `excludedAtDirectives`: List `@`-prefixed directives to ignore as header candidates to keep `@ts-nocheck`, `@format`, `@jest-environment`, and similar pragmas out of the cursor header. Set to `[]` to treat all `@` lines as candidates.

## Examples

Examples of **incorrect** code for this rule:

```ts
// No header at all
import { HttpsError } from '../errors/HttpsError';
export const validateUser = (userId: string) => !!userId;
```

```ts
/**
 * @fileoverview User authentication utilities
 */

/**
 * @fileoverview User authentication utilities
 */
export const validateUser = (userId: string) => !!userId;
```

```ts
import { utils } from './utils';

/**
 * @fileoverview Header after imports is ignored
 */
export const lateHeader = () => utils;
```

Examples of **correct** code for this rule:

```ts
/**
 * @fileoverview User authentication utilities
 * @author BluMint
 */
import { HttpsError } from '../errors/HttpsError';
export const validateUser = (userId: string) => !!userId;
```

```ts
// allowSplitHeaders = true (default)
/**
 * @fileoverview User authentication utilities
 */
/**
 * @author BluMint
 */
export const helper = () => {};
```

```ts
// allowSplitHeaders = false â†’ requires single block
/**
 * @fileoverview User authentication utilities
 * @author BluMint
 */
export const helper = () => {};
```

## When Not To Use It

Disable this rule if your project does not require cursor headers, or if another tool guarantees file-level metadata and you want to allow multiple header blocks or omit `@fileoverview`.
