# Enforce naming conventions for React types (`@blumintinc/blumint/enforce-react-type-naming`)

ğŸ’¼ This rule is enabled in the âœ… `recommended` config.

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

ğŸ’¼ This rule is enabled in the âœ… `recommended` config.

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

ğŸ’¼ This rule is enabled in the âœ… `recommended` config.

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

This rule enforces naming conventions for variables and parameters based on their React-related type annotations:

- Variables and parameters of type `ReactNode` or `JSX.Element` must have **lowercase** names.
- Variables and parameters of type `ComponentType` or `FC` (FunctionComponent) must have **uppercase** names.

This ensures consistency with React's conventions, improving readability and maintainability. By distinguishing between rendered elements (which should be treated as values) and component definitions (which are typically PascalCase), the rule helps prevent common mistakes and aligns with best practices.

## Rule Details

Examples of **incorrect** code for this rule:

```tsx
const MyComponent: JSX.Element = <div>Hello</div>; // âŒ Should be lowercase
const element: ComponentType = () => <div />; // âŒ Should be uppercase

function useCustomHook(Component: ReactNode) { // âŒ Should be lowercase
  return <Component />;
}
```

Examples of **correct** code for this rule:

```tsx
const myComponent: JSX.Element = <div>Hello</div>; // âœ… Lowercase for JSX.Element
const Element: ComponentType = () => <div />; // âœ… Uppercase for ComponentType

function useCustomHook(component: ReactNode) { // âœ… Lowercase for ReactNode
  return <component />;
}
```

## When Not To Use It

You might want to disable this rule if your project follows different naming conventions for React components and elements.

## Further Reading

- [React Components and Elements](https://reactjs.org/blog/2015/12/18/react-components-elements-and-instances.html)
