# Discourage using margin properties (margin, marginLeft, marginRight, marginTop, marginBottom, mx, my, etc.) for spacing in MUI components. Instead, prefer defining spacing with padding, gap, or the spacing prop for more predictable layouts (`@blumintinc/blumint/no-margin-properties`)

üíº This rule is enabled in the ‚úÖ `recommended` config.

<!-- end auto-generated rule header -->

## Rule Details

This rule discourages the use of margin properties in MUI components to promote more predictable layouts. Instead of using margin properties, prefer using padding, gap, or the spacing prop.

## Options

This rule accepts an options object with the following properties:

- `autofix` (boolean, default: `false`): Controls whether the rule should provide automatic fixes. Currently, no autofix functionality is implemented, but this option is available for future use.

### Default Configuration

```json
{
  "@blumintinc/blumint/no-margin-properties": "warn"
}
```

### Custom Configuration

```json
{
  "@blumintinc/blumint/no-margin-properties": ["warn", { "autofix": false }]
}
```

## Examples

### ‚ùå Incorrect

```jsx
// Using margin properties in sx prop
<Box sx={{ margin: 2, marginTop: 3 }} />

// Using margin shorthand properties
<Stack sx={{ mx: 2, my: 1 }} />

// Using margin as direct props
<Box margin={2} marginTop={3} />
```

### Exceptions / When Not To Use It

You may want to disable this rule for:
- Third-party layout components that require margin props.
- Legacy components that do not support the recommended spacing API.
- Cases where margins are the only viable option.

In such cases, prefer adding an `eslint-disable` comment with a brief explanation.

### ‚úÖ Correct

```jsx
// Using padding instead
<Box sx={{ padding: 2, paddingTop: 3 }} />

// Using spacing prop for Stack
<Stack spacing={2} />

// Using gap for layout
<Box sx={{ display: 'flex', gap: 2 }} />
```
