# Disallow using uuidv4Base62() to generate keys for elements in a list or loop (`@blumintinc/blumint/no-uuidv4-base62-as-key`)

ðŸ’¼ This rule is enabled in the âœ… `recommended` config.

<!-- end auto-generated rule header -->

Avoid using `uuidv4Base62()` as React list keys because each call returns a new value, breaking React's ability to track list items.

## Rule Details

React reconciliation depends on stable keys to preserve component identity. Keys generated by `uuidv4Base62()` change on every render, so React treats each element as new: components unmount and remount, DOM nodes churn, and local state or focus can disappear. Random keys also hide ordering bugs because React cannot match items between renders. This rule flags keys that come from `uuidv4Base62()` anywhere inside list or loop rendering.

### How to fix

- Use a stable identifier from the data: database id, slug, or another persistent field.
- If you must synthesize keys, compute them once and reuse them (for example, memoize an array of ids for a dataset) rather than generating a new UUID during render.
- For small, static arrays, an index key is safer than regenerating UUIDs each render, though a domain identifier is preferred.

### Examples

#### âŒ Incorrect

```jsx
{items.map((item) => (
  <div key={uuidv4Base62()}>{item.name}</div>
))}

const itemKeys = items.map((item) => ({ ...item, key: uuidv4Base62() }));
{itemKeys.map((item) => (
  <div key={item.key}>{item.name}</div>
))}
```

#### âœ… Correct

```jsx
{items.map((item) => (
  <div key={item.id}>{item.name}</div>
))}

{items.map((item, index) => (
  <div key={`${item.name}-${index}`}>{item.name}</div>
))}

function Component({ items }) {
  const memoizedIds = React.useMemo(
    () => items.map((_item) => uuidv4Base62()),
    [items],
  );
  return items.map((item, index) => (
    <div key={memoizedIds[index]}>{item.name}</div>
  ));
}
```

## When Not To Use It

Disable this rule only if you intentionally need components to remount on every render and accept the lost state and DOM churn that results. In most cases a stable domain identifier or memoized key set is the safer choice.

## Further Reading

- [React Lists and Keys](https://reactjs.org/docs/lists-and-keys.html)
- [React Reconciliation](https://reactjs.org/docs/reconciliation.html#keys)
- [Understanding React's key prop](https://kentcdodds.com/blog/understanding-reacts-key-prop)
