# Enforce using setAll() and overwriteAll() instead of multiple set() or overwrite() calls (`@blumintinc/blumint/prefer-batch-operations`)

ðŸ’¼ This rule is enabled in the âœ… `recommended` config.

ðŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

Repeated `DocSetter.set()` or `DocSetter.overwrite()` calls inside a loop, array callback, or `Promise.all()` create one Firestore write per document. That pattern slows down writes and can leave partial updates if any later call fails. `DocSetter.setAll()` and `DocSetter.overwriteAll()` batch the same documents into one grouped write with predictable latency and fewer network round trips.

When this rule fires, the lint message tells you where the repetition happens (for example, `for...of loop`, `map() callback`, or `Promise.all()`) and points to the matching batch method as the fix.

## Rule Details

- Flags multiple `DocSetter.set()` or `DocSetter.overwrite()` calls inside `for`/`for...of`/`for...in`/`while`/`do...while`, `forEach`/`map`/`filter`/`reduce` callbacks, and `Promise.all()` arrays.
- Skips mixed operations that cannot be batched together (for example, mixing `set()` with unrelated service calls) and multiple setter instances in the same loop.
- Ignores `Map.set()` because it is not a Firestore write.
- Single `set()`/`overwrite()` calls remain valid; the rule only cares about repeated calls that should be batched.

## Examples

### âŒ Incorrect

```ts
const setter = new DocSetter(collectionRef);

await Promise.all([
  setter.set(doc1),
  setter.set(doc2),
  setter.set(doc3),
]);
```

```ts
const setter = new DocSetter(collectionRef);

for (const doc of documents) {
  await setter.overwrite(doc);
}
```

### âœ… Correct

```ts
const setter = new DocSetter(collectionRef);

await setter.setAll([doc1, doc2, doc3]);
```

```ts
const setter = new DocSetter(collectionRef);
const docsNeedingOverwrite = documents.filter((doc) => doc.shouldOverwrite);

await setter.overwriteAll(docsNeedingOverwrite);
```

### âœ… Allowed patterns

```ts
const setter = new DocSetter(collectionRef);

await Promise.all([
  setter.set(doc),
  sendEmail(), // different operation type
]);
```

```ts
const seen = new Map();
items.forEach((item) => seen.set(item.id, true)); // Map.set() is allowed
```
